stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -605,44 +673,73 @@ impl NakamotoBlock {
             return None;
         }
 
-        if coinbase_positions.len() > 1 {
-            // has more than one coinbase
+        if coinbase_positions.len() > 1 || tenure_change_positions.len() > 1 {
+            // never valid to have more than one of each
             return Some(false);
         }
 
         if coinbase_positions.len() == 1 && tenure_change_positions.len() == 0 {
-            // has a coinbase but no tenure change
+            // coinbase unaccompanied by a tenure change
             return Some(false);
         }
 
-        if coinbase_positions.len() == 0 && tenure_change_positions.len() > 0 {
-            // has tenure-changes but no coinbase
-            return Some(false);
-        }
+        if coinbase_positions.len() == 0 && tenure_change_positions.len() == 1 {
+            // this is possibly a block with a tenure-extend transaction.
+            // It must be the first tx
+            if tenure_change_positions != vec![0] {
+                // wrong position
+                return Some(false);
+            }
+
+            // must be a non-sortition-triggered tenure change
+            let TransactionPayload::TenureChange(tc_payload) = &self.txs[0].payload else {
+                // this transaction is not a tenure change
+                // (should be unreachable)
+                return Some(false);
+            };
 
-        // tenure-changes must all come first, and must be in order
-        for (i, pos) in tenure_change_positions.iter().enumerate() {
-            if &i != pos {
-                // tenure-change is out of place
+            if tc_payload.cause.expects_sortition() {
+                // not valid
                 return Some(false);
             }
+
+            // not a tenure-start block, but syntactically valid w.r.t. tenure changes
+            return None;
```

Comment:
Re (2), the consensus hash of a tenure-extension _must be the same as_ the last winning block-commit's sortition's consensus hash.  This is true for all tenure-changes -- we identify tenures by their block-commit's sortition's consensus hashes, and the `consensus_hash` field in a tenure change binds it to that tenure.

The reason for this is because calculating block rewards and maturity becomes really tricky otherwise, and moreover, does not provide the same incentives.  Namely, we want miners to get rewarded only when the _tenure_ gets sufficiently confirmed by subsequent tenures -- we want them to keep working on the chain in order to get paid.  But to do this, we'd only count a reward as matured if there have been enough sortition-induced tenure changes.

Reply:
> the consensus hash of a tenure-extension _must be the same as_ the last winning block-commit's sortition's consensus hash. This is true for all tenure-changes -- we identify tenures by their block-commit's sortition's consensus hashes, and the `consensus_hash` field in a tenure change binds it to that tenure.

There are two consensus hash fields in the tenure changes, right? The “previous consensus hash” must match the prior tenure, but the new consensus hash does not always — in the case of a sortition, it would never, but in extensions due to a missed sortition as well. In that last case, referencing such as a single tenure (i.e., burn block 1 _and_ the extended to burn block 2) becomes tricky, because all the blocks of the tenure do not have the same consensus hash.

> The reason for this is because calculating block rewards and maturity becomes really tricky otherwise, and moreover, does not provide the same incentives.

I mostly agree with this: miner rewards and maturity shouldn’t “tick” on non-sortition tenure change txs. However, I think that if the implementation treats every tenure change transaction as a new tenure, this property can still be implemented (my suggestion was to essentially count coinbase transactions, which are only allowed when a tenure change occurs due to sortition).

Reply:
> because all the blocks of the tenure do not have the same consensus hash.

This is no longer true.  The `consensus_hash` field in `NakamotoBlockHeader` identifies the tenure, and the tenure is defined as the sequence of blocks produced by the last sortition-selected miner.  Subsequent Bitcoin blocks and failed sortitions can happen, but the consensus hash of the ongoing stream of Nakamoto blocks will remain the same as that of the sortition which selected the currently-active miner.

This differs from the SIP, btw.  I need to update the SIP.

Reply:
How do extended blocks pick up the new burn chain changes, then?

Reply:
I'm in the process of tweaking the `TenureChangePayload` and `tenure_headers` table to track three consensus hashes:

* `tenure_consensus_hash` -- the consensus hash that identifies the winning miner (i.e. it's the consensus hash of the sortition in which the winning miner was selected)
* `prev_tenure_consensus_hash` -- the consensus hash of the previous winning miner.  It's not guaranteed to be the parent sortition consensus hash of `tenure_consensus_hash`, since the miner's tenure may span multiple sortitions
* `sortition_consensus_hash` -- the consensus hash of the sortition which triggered Stackers to produce this `TenureChange` transaction

The `sortition_consensus_hash` will identify new burnchain changes.  However, it only does so when there is an active miner.  As such, the `sortition_consensus_hash` values in `nakamoto_tenures` don't correspond to sortitions 1-to-1 -- there can be sortitions that do not have a corresponding `nakamoto_tenures` row, because it's possible that there were no active miners during a sortition.

What I expect the Clarity VM to do (specifically, the `BurnStateDB` implementation) to handle `get-burn-block-info?` is to perform an ancestor query from the highest-seen `sortition_consensus_hash`'s sortition back to the block hash given in `get-burn-block-info?` in order to (1) determine that the block exists and (2) identify where that block falls in the burnchain history.

Reply:
So you are saying that if a sortition doesn’t occur, then new burn chain operations will not be processed until the next sortition occurs.

What is the advantage of treating tenures this way?

Reply:
Take a look at what I have now in `src/chainstate/nakamoto/tenure.rs`